packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +661,139 @@ export class RunningSummarizer implements IDisposable {
 		});
 	}
 
+	/**
+	 * Tries to summarize with dynamic retries, i.e., retry is based on the failure params and error received.
+	 * For instance, if summarize failed due to a server error that has `retryAfterSeconds` set, wait for that duration
+	 * and retry summarize. Similarly, if summary generation failed but the error is retriable, retry before bailing out.
+	 */
+	private async trySummarizeWithDynamicRetries(
+		reason: SummarizeReason,
+		cancellationToken = this.cancellationToken,
+	) {
+		// Currently, we try 2 overall attempts with retries in each attempt. The second attempt will close the
+		// summarizer at the summarization call site but we need to keep this code for the time being. This will be
+		// reduced to a single attempt.
+		const attemptOptions: ISummarizeOptions[] = [
+			{ refreshLatestAck: false, fullTree: false },
+			{ refreshLatestAck: true, fullTree: false },
+		];
+		for (const summarizeOptions of attemptOptions) {
+			let previousStage: SummaryStage | undefined;
+
+			// The maximum number of retries across failure stages. If summarization does not succeed by these number
+			// of retries, quit and close the summarizer.
+			const maxRetries = defaultMaxRetries;
+			let currentRetries = -1;
+
+			// The maximum number of retries for a given stage. The default value is 1 but this can be set via failure
+			// params. Each failure is different and some can be succeeded by retrying more often than others.
+			// If summarization fails at different stages during retries, each stage gets some retries. However, the
+			// total number of retries across stages cannot exceed `maxRetries`.
+			let maxRetriesForStage = 1;
+			let currentRetriesForStage = -1;
```

Comment:
I'm surprised this starts at -1.

Reply:
Yeah, it's weird. I can change it to 0 based.

Reply:
Changed it to 0 based. Also renamed `retries` to `attempts` to make things clearer.